docs/magento-layout-loader.md

```diff
@@ -0,0 +1,288 @@
+                render={() => <div>This content will be removed</div>}
+            />
+        </div>
+    );
+}
+```
+
+##### Output
+
+```js
+import React from 'react';
+
+function render() {
+    return (
+        <div className="wrapper" data-mid="any.container.id">
+            The container below will be removed
+        </div>
+    );
+}
+```
+
+</details>
+
+#### insertBefore
+
+##### Configuration
+
+```js
+{
+    "operation": "insertBefore",
+    "targetContainer": "any.container.id",
+    "targetChild": "container.child.id",
+    "componentPath": "/Absolute/path/to/a/component.js"
+}
+```
+
+<details>
+<summary>Example</summary>
+
+##### Input
+
+```js
+import React from 'react';
+import { ContainerChild } from '@magento/peregrine';
+
+function render() {
+    return (
+        <div className="wrapper" data-mid="any.container.id">
+            <ContainerChild
+                id="container.child.id"
+                render={() => <div>Some Content</div>}
+            />
+        </div>
+    );
+}
+```
+
+##### Output
+
+```js
+import React from 'react';
+import { ContainerChild } from '@magento/peregrine';
+import _Extension from '/Absolute/path/to/a/component.js';
+
+function render() {
+    return (
+        <div className="wrapper" data-mid="any.container.id">
+            <_Extension />
+            <ContainerChild
+                id="container.child.id"
+                render={() => <div>Some Content</div>}
+            />
+        </div>
+    );
+}
+```
+
+</details>
+
+#### insertAfter
+
+##### Configuration
+
+```js
+{
+    "operation": "insertAfter",
+    "targetContainer": "any.container.id",
+    "targetChild": "container.child.id",
+    "componentPath": "/Absolute/path/to/a/component.js"
+}
+```
+
+<details>
+<summary>Example</summary>
+
+##### Input
+
+```js
+import React from 'react';
+import { ContainerChild } from '@magento/peregrine';
+
+function render() {
+    return (
+        <div className="wrapper" data-mid="any.container.id">
+            <ContainerChild
+                id="container.child.id"
+                render={() => <div>Some Content</div>}
+            />
+        </div>
+    );
+}
+```
+
+##### Output
+
+```js
+import React from 'react';
+import { ContainerChild } from '@magento/peregrine';
+import _Extension from '/Absolute/path/to/a/component.js';
+
+function render() {
+    return (
+        <div className="wrapper" data-mid="any.container.id">
+            <ContainerChild
+                id="container.child.id"
+                render={() => <div>Some Content</div>}
+            />
+            <_Extension />
+        </div>
+    );
+}
+```
+
+</details>
+
+## FAQ
+
+### How are the configurations for individual extensions/modules collected and provided to the loader?
+
+This functionality has not been completed yet. There is outstanding work to be
+done in Magento 2 to collect and expose aggregated configuration from each
+module.
+
+### Why is a `Container` required to be a DOM Element?
```

Comment:
As much as I understand your reasons here, I think it's a pretty heavy drawback.
Especially when you want to avoid to use DOM elements on top level components (and since extensions are more likely to accurd on top level components... :s).

What about creating a `Container` component directly ? It could be defaulted to a `div` and have a render prop that let the user decide whether it's a div, a span or whatsoever. IMO, it makes it even more explicit that it's an extension point.

Reply:
The original proposal was for a `Container` composite component, but a few of us didn't like the ergonomics of it when actually trying it out in some demo apps. We plan to start building more of our base theme on this, and we'll hopefully find the ideal API as we (whether it's this, `Container`, or something else) as we continue.

Luckily, with 0 production apps relying on this, nothing is set in stone.